In [2]:
from pathlib import Path

DATA_DIR = Path(
    r"C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행"
)

# 01. 하위 폴더 포함 전체 파일 탐색
all_files = [p for p in DATA_DIR.rglob("*") if p.is_file()]

print("전체 파일 수:", len(all_files))

# 02. 확장자 종류 확인
extensions = {}

for f in all_files:
    ext = f.suffix.lower()
    extensions[ext] = extensions.get(ext, 0) + 1

print("\n확장자별 파일 수:")
print(extensions)

# 03. JSON 파일 찾기
json_files = [f for f in all_files if f.suffix.lower() == ".json"]

print("\nJSON 파일 수:", len(json_files))

print("\nJSON 샘플 경로:")
for f in json_files[:10]:
    print(f)

전체 파일 수: 26572

확장자별 파일 수:
{'.json': 26572}

JSON 파일 수: 26572

JSON 샘플 경로:
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000029.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000075.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000085.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000102.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000197.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000201.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000212.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방데이터\2.데이터(NIA)\Training\01.원천데이터\TS_은행\01\21-1_bk_01_000297.json
C:\Users\user\Downloads\25.금융분야 고객상담 데이터\3.개방

In [3]:
files = [p for p in DATA_DIR.rglob("*") if p.is_file()]

print("전체 파일 수:", len(files))

for f in files[:20]:
    print(f.name)

전체 파일 수: 26572
21-1_bk_01_000029.json
21-1_bk_01_000075.json
21-1_bk_01_000085.json
21-1_bk_01_000102.json
21-1_bk_01_000197.json
21-1_bk_01_000201.json
21-1_bk_01_000212.json
21-1_bk_01_000297.json
21-1_bk_01_000309.json
21-1_bk_01_000312.json
21-1_bk_01_000314.json
21-1_bk_01_000320.json
21-1_bk_01_000359.json
21-1_bk_01_000362.json
21-1_bk_01_000379.json
21-1_bk_01_000402.json
21-1_bk_01_000409.json
21-1_bk_01_000415.json
21-1_bk_01_000423.json
21-1_bk_01_000437.json


In [4]:
import json
from pprint import pprint

# 01. 첫 번째 JSON 열기
sample_path = files[0]

with open(sample_path, "r", encoding="utf-8-sig") as f:
    sample = json.load(f)

print("파일명:", sample_path.name)
print("데이터 타입:", type(sample))

# 02. 최상위 구조 확인
if isinstance(sample, dict):
    print("\n최상위 KEY:")
    print(list(sample.keys()))

elif isinstance(sample, list):
    print("\n리스트 길이:", len(sample))
    if sample:
        print("첫 번째 원소 타입:", type(sample[0]))

# 03. 실제 내용 확인
print("\n원본 내용:")
pprint(sample, depth=5, width=120)

파일명: 21-1_bk_01_000029.json
데이터 타입: <class 'dict'>

최상위 KEY:
['source', 'consulting']

원본 내용:
{'consulting': {'consulting_category': '은행', 'consulting_topic': '거래내역/잔액조회'},
 'source': {'client_age': '50~59세',
            'client_gender': '남',
            'consulting_client': '기존고객',
            'consulting_client_type': '개인사업자',
            'consulting_content': 'TX 안녕하세요 OO은행 고객센터입니다. 오늘 무엇을 도와드릴까요?\n'
                                  'RX 안녕하세요 수고 많으십니다. 저는 현재 입주 계약과 관련된 대출 계약금 입금 내역에 대해 문의드리고 싶습니다.\n'
                                  'TX 네 고객님 말씀을 자세히 들어보겠습니다. 어떤 부분이 궁금하신가요?\n'
                                  'RX 제가 대부계 계약금을 ●● ● ●원 정도 보내드렸는데 제 통장에 표시된 금액이 ●●● ●원씩 차이가 나서 확인이 필요합니다. 이 차이가 발생한 '
                                  '이유가 무엇인지 알려주실 수 있나요?\n'
                                  'TX 확인해 보니 시스템 상에서 입금 금액과 실제 통장에 반영된 금액 사이에 일시적인 연동 오류가 발생할 수 있다고 안내드립니다. 이 오류는 보통 '
                                  '영업시간 이후에 자동 정정됩니다.\n'
                                  'RX 그렇다면 오류가 자동으로 정정되는

In [5]:
import json
from collections import Counter

source_keys = Counter()
consulting_keys = Counter()
errors = []

for f in files:
    try:
        with open(f, "r", encoding="utf-8-sig") as fp:
            data = json.load(fp)

        source = data.get("source", {})
        consulting = data.get("consulting", {})

        for key in source.keys():
            source_keys[key] += 1

        for key in consulting.keys():
            consulting_keys[key] += 1

    except Exception as e:
        errors.append((f.name, str(e)))

print("전체 JSON:", len(files))

print("\n[source 컬럼]")
for k, v in source_keys.items():
    print(f"{k}: {v:,}")

print("\n[consulting 컬럼]")
for k, v in consulting_keys.items():
    print(f"{k}: {v:,}")

print("\n읽기 오류:", len(errors))

전체 JSON: 26572

[source 컬럼]
source_institution: 26,572
source_id: 26,572
source_date: 26,572
client_gender: 26,572
client_age: 26,572
consulting_client: 26,572
consulting_client_type: 26,572
source_length: 26,572
consulting_content: 26,572

[consulting 컬럼]
consulting_category: 26,572
consulting_topic: 26,572

읽기 오류: 0


In [6]:
import json
import pandas as pd

rows = []

for f in files:
    with open(f, "r", encoding="utf-8-sig") as fp:
        data = json.load(fp)

    row = {
        "file_name": f.name,

        # source
        "source_institution": data["source"]["source_institution"],
        "source_id": data["source"]["source_id"],
        "source_date": data["source"]["source_date"],
        "client_gender": data["source"]["client_gender"],
        "client_age": data["source"]["client_age"],
        "consulting_client": data["source"]["consulting_client"],
        "consulting_client_type": data["source"]["consulting_client_type"],
        "source_length": data["source"]["source_length"],
        "consulting_content": data["source"]["consulting_content"],

        # consulting
        "consulting_category": data["consulting"]["consulting_category"],
        "consulting_topic": data["consulting"]["consulting_topic"],
    }

    rows.append(row)

df = pd.DataFrame(rows)

print("shape:", df.shape)
display(df.head())

shape: (26572, 12)


,file_name,source_institution,source_id,source_date,client_gender,client_age,consulting_client,consulting_client_type,source_length,consulting_content,consulting_category,consulting_topic
0,21-1_bk_01_000029.json,하나은행,21-1_bk_01_000029,202506,남,50~59세,기존고객,개인사업자,1331,TX 안녕하세요 OO은행 고객센터입니다. 오늘 무엇을 도와드릴까요?\nRX 안녕하세...,은행,거래내역/잔액조회
1,21-1_bk_01_000075.json,하나은행,21-1_bk_01_000075,202505,여,40~49세,기존고객,None,1119,"고객은 최근 출금 거래의 ATM 위치와 수수료면제 여부, 해당계좌의 무단 사용 여부...",은행,거래내역/잔액조회
2,21-1_bk_01_000085.json,하나은행,21-1_bk_01_000085,202506,여,60~69세,기존고객,None,1727,고객은 외화 계좌를 원화 계좌로 전환하고 카드 발급을 위해 필요한 절차와 최소 예치...,은행,거래내역/잔액조회
3,21-1_bk_01_000102.json,하나은행,21-1_bk_01_000102,202505,여,60~69세,기존고객,개인사업자,979,"고객은 오늘 입금된 내역과 금액을 확인하고 싶다고 문의했고, 상담원은 계좌 확인 후...",은행,거래내역/잔액조회
4,21-1_bk_01_000197.json,하나은행,21-1_bk_01_000197,202506,여,30~39세,기존고객,None,1448,"고객은 최근 이체가 입금되지 않아 확인을 요청하고, 거래 중지 계좌의 자동이체 가능...",은행,거래내역/잔액조회


In [7]:
# 01. 데이터 타입
print(df.dtypes)

# 02. 결측치
display(
    df.isnull()
      .sum()
      .sort_values(ascending=False)
      .to_frame("missing")
)

# 03. 완전 중복
print("완전 중복:", df.duplicated().sum())

# 04. source_id 중복
print("source_id 중복:", df["source_id"].duplicated().sum())

# 05. 컬럼별 고유값 수
display(
    df.nunique()
      .sort_values()
      .to_frame("unique_count")
)

file_name                 object
source_institution        object
source_id                 object
source_date                int64
client_gender             object
client_age                object
consulting_client         object
consulting_client_type    object
source_length              int64
consulting_content        object
consulting_category       object
consulting_topic          object
dtype: object


,missing
consulting_client_type,19115
file_name,0
source_id,0
source_institution,0
source_date,0
client_gender,0
client_age,0
consulting_client,0
source_length,0
consulting_content,0


완전 중복: 0
source_id 중복: 0


,unique_count
source_institution,1
consulting_client_type,1
consulting_category,1
client_gender,2
consulting_client,2
source_date,2
client_age,7
consulting_topic,9
source_length,1962
consulting_content,26562


In [8]:
cols = [
    "source_institution",
    "client_gender",
    "client_age",
    "consulting_client",
    "consulting_client_type",
    "consulting_category",
]

for col in cols:
    print(f"\n[{col}]")
    print(df[col].value_counts(dropna=False))


[source_institution]
source_institution
하나은행    26572
Name: count, dtype: int64

[client_gender]
client_gender
남    14039
여    12533
Name: count, dtype: int64

[client_age]
client_age
50~59세    6998
40~49세    5895
60~69세    4842
30~39세    4667
20~29세    2041
70~79세    1784
80세이상      345
Name: count, dtype: int64

[consulting_client]
consulting_client
기존고객    26342
신규고객      230
Name: count, dtype: int64

[consulting_client_type]
consulting_client_type
None     19115
개인사업자     7457
Name: count, dtype: int64

[consulting_category]
consulting_category
은행    26572
Name: count, dtype: int64


In [9]:
print("상담 Topic 종류:", df["consulting_topic"].nunique())

display(
    df["consulting_topic"]
      .value_counts()
      .head(30)
      .to_frame("count")
)

상담 Topic 종류: 9


,count
consulting_topic,
대출문의(만기/연장/조회등),7946
이자/연체금액,5786
거래내역/잔액조회,2675
자동이체조회,2657
금융거래한도/비대면한도계좌,1994
"만기,연장/해지,수신",1914
환전문의,1443
중계요청/착오송금,1157
부수거래금리감면,1000


In [10]:
# 01. 메타데이터 실제 값 확인
check_cols = [
    "source_institution",
    "source_date",
    "client_gender",
    "client_age",
    "consulting_client",
    "consulting_client_type",
    "consulting_category",
    "consulting_topic"
]

for col in check_cols:
    print(f"\n[{col}]")
    print(df[col].value_counts(dropna=False))


[source_institution]
source_institution
하나은행    26572
Name: count, dtype: int64

[source_date]
source_date
202506    13830
202505    12742
Name: count, dtype: int64

[client_gender]
client_gender
남    14039
여    12533
Name: count, dtype: int64

[client_age]
client_age
50~59세    6998
40~49세    5895
60~69세    4842
30~39세    4667
20~29세    2041
70~79세    1784
80세이상      345
Name: count, dtype: int64

[consulting_client]
consulting_client
기존고객    26342
신규고객      230
Name: count, dtype: int64

[consulting_client_type]
consulting_client_type
None     19115
개인사업자     7457
Name: count, dtype: int64

[consulting_category]
consulting_category
은행    26572
Name: count, dtype: int64

[consulting_topic]
consulting_topic
대출문의(만기/연장/조회등)    7946
이자/연체금액            5786
거래내역/잔액조회          2675
자동이체조회             2657
금융거래한도/비대면한도계좌     1994
만기,연장/해지,수신        1914
환전문의               1443
중계요청/착오송금          1157
부수거래금리감면           1000
Name: count, dtype: int64


In [11]:
# 02. 동일 상담내용 중복 확인
dup_content = df[
    df.duplicated(subset=["consulting_content"], keep=False)
].sort_values("consulting_content")

print("중복 상담 행 수:", len(dup_content))
print("중복된 상담문 종류:", dup_content["consulting_content"].nunique())

display(
    dup_content[
        [
            "source_id",
            "source_date",
            "client_gender",
            "client_age",
            "consulting_topic",
            "consulting_content"
        ]
    ]
)

중복 상담 행 수: 13
중복된 상담문 종류: 3


,source_id,source_date,client_gender,client_age,consulting_topic,consulting_content
8101,21-1_bk_04_024722,202506,여,50~59세,"만기,연장/해지,수신",
11105,21-1_bk_06_003660,202505,여,30~39세,이자/연체금액,
14774,21-1_bk_06_022224,202505,여,50~59세,이자/연체금액,
15266,21-1_bk_06_024547,202505,여,50~59세,이자/연체금액,
16124,21-1_bk_06_028453,202505,남,40~49세,이자/연체금액,
16915,21-1_bk_07_021368,202505,여,40~49세,부수거래금리감면,
22602,21-1_bk_08_019992,202506,여,50~59세,대출문의(만기/연장/조회등),
23720,21-1_bk_08_023668,202506,여,60~69세,대출문의(만기/연장/조회등),
26302,21-1_bk_09_024026,202506,여,50~59세,환전문의,
17662,21-1_bk_08_001857,202506,여,60~69세,대출문의(만기/연장/조회등),"고객은 대출 연장 절차, 필요 서류, 신청 방법, 금리 변동, 상환 일정, 조기 상..."


In [12]:
# 03. 상담 텍스트 길이 실제값 계산
df["text_length"] = df["consulting_content"].str.len()

display(df["text_length"].describe())

print("빈 문자열:", (df["consulting_content"].str.strip() == "").sum())

print(
    "source_length와 실제 글자수 동일:",
    (df["source_length"] == df["text_length"]).sum(),
    "/",
    len(df)
)

count    26572.000000
mean       589.310590
std        725.967841
min          0.000000
25%        149.000000
50%        188.000000
75%       1082.250000
max       4188.000000
Name: text_length, dtype: float64

빈 문자열: 9
source_length와 실제 글자수 동일: 9 / 26572


In [13]:
# 01. consulting_client_type 결측 행 확인
missing_client_type = df[df["consulting_client_type"].isna()].copy()

print("결측 행 수:", len(missing_client_type))

display(
    missing_client_type[
        [
            "source_id",
            "client_gender",
            "client_age",
            "consulting_client",
            "consulting_client_type",
            "consulting_topic",
            "consulting_content"
        ]
    ].head(30)
)

결측 행 수: 19115


,source_id,client_gender,client_age,consulting_client,consulting_client_type,consulting_topic,consulting_content
1,21-1_bk_01_000075,여,40~49세,기존고객,None,거래내역/잔액조회,"고객은 최근 출금 거래의 ATM 위치와 수수료면제 여부, 해당계좌의 무단 사용 여부..."
2,21-1_bk_01_000085,여,60~69세,기존고객,None,거래내역/잔액조회,고객은 외화 계좌를 원화 계좌로 전환하고 카드 발급을 위해 필요한 절차와 최소 예치...
4,21-1_bk_01_000197,여,30~39세,기존고객,None,거래내역/잔액조회,"고객은 최근 이체가 입금되지 않아 확인을 요청하고, 거래 중지 계좌의 자동이체 가능..."
5,21-1_bk_01_000201,여,30~39세,기존고객,None,거래내역/잔액조회,"RX 안녕하세요, 저는 ■■■이고 현재 ○○은행에 거래하고 있는 고객입니다. 오늘 ..."
7,21-1_bk_01_000297,남,60~69세,기존고객,None,거래내역/잔액조회,"TX 안녕하세요, OO은행입니다. 무엇을 도와드릴까요? 고객님께서 거래내역서를 원하..."
8,21-1_bk_01_000309,여,20~29세,기존고객,None,거래내역/잔액조회,TX 안녕하세요 OO은행 고객센터입니다. 무엇을 도와드릴까요?\nRX 안녕하세요. ...
9,21-1_bk_01_000312,여,50~59세,기존고객,None,거래내역/잔액조회,"TX 안녕하세요, 고객님. 무엇을 도와드릴까요?\nRX 안녕하세요. 제 통장에 자동..."
10,21-1_bk_01_000314,남,40~49세,기존고객,None,거래내역/잔액조회,"고객은 이체가 되지 않는 원인과 오류 코드 의미, 재인증 방법, 계좌 이체 제한 확..."
12,21-1_bk_01_000359,여,20~29세,기존고객,None,거래내역/잔액조회,"RX 안녕하세요. 저는 ●●● 계좌에 입금된 내역이 있는지 확인하고 싶은데요, 입금..."
14,21-1_bk_01_000379,여,40~49세,기존고객,None,거래내역/잔액조회,"TX 안녕하세요, OO은행입니다. 전화를 주셔서 무엇을 도와드릴까요?\nRX 안녕하..."


In [14]:
# 01. 불필요 컬럼 제거
df = df.drop(columns=["consulting_client_type"])

print(df.shape)
print(df.columns.tolist())

(26572, 12)
['file_name', 'source_institution', 'source_id', 'source_date', 'client_gender', 'client_age', 'consulting_client', 'source_length', 'consulting_content', 'consulting_category', 'consulting_topic', 'text_length']


In [15]:
# 02. 결측 재확인
display(
    df.isnull()
      .sum()
      .sort_values(ascending=False)
      .to_frame("missing")
)

,missing
file_name,0
source_institution,0
source_id,0
source_date,0
client_gender,0
client_age,0
consulting_client,0
source_length,0
consulting_content,0
consulting_category,0


In [16]:
# 01. 완전 중복
print("완전 중복:", df.duplicated().sum())

# 02. source_id 중복
print("source_id 중복:", df["source_id"].duplicated().sum())

# 03. 상담내용 중복
print("consulting_content 중복:", df["consulting_content"].duplicated().sum())

완전 중복: 0
source_id 중복: 0
consulting_content 중복: 10


In [17]:
dup_content = df[
    df.duplicated(subset=["consulting_content"], keep=False)
].sort_values("consulting_content")

print("중복 포함 행 수:", len(dup_content))

display(
    dup_content[
        [
            "source_id",
            "source_date",
            "client_gender",
            "client_age",
            "consulting_topic",
            "consulting_content"
        ]
    ]
)

중복 포함 행 수: 13


,source_id,source_date,client_gender,client_age,consulting_topic,consulting_content
8101,21-1_bk_04_024722,202506,여,50~59세,"만기,연장/해지,수신",
11105,21-1_bk_06_003660,202505,여,30~39세,이자/연체금액,
14774,21-1_bk_06_022224,202505,여,50~59세,이자/연체금액,
15266,21-1_bk_06_024547,202505,여,50~59세,이자/연체금액,
16124,21-1_bk_06_028453,202505,남,40~49세,이자/연체금액,
16915,21-1_bk_07_021368,202505,여,40~49세,부수거래금리감면,
22602,21-1_bk_08_019992,202506,여,50~59세,대출문의(만기/연장/조회등),
23720,21-1_bk_08_023668,202506,여,60~69세,대출문의(만기/연장/조회등),
26302,21-1_bk_09_024026,202506,여,50~59세,환전문의,
17662,21-1_bk_08_001857,202506,여,60~69세,대출문의(만기/연장/조회등),"고객은 대출 연장 절차, 필요 서류, 신청 방법, 금리 변동, 상환 일정, 조기 상..."


In [18]:
# 05. 상담내용이 없는 행 제거
# 전체 대비 비중을 함께 확인한 뒤 제거한다.

total_count = len(df)

empty_mask = (
    df["consulting_content"].isna()
    | df["consulting_content"].str.strip().eq("")
)

empty_count = empty_mask.sum()
empty_ratio = empty_count / total_count * 100

print(f"전체 데이터: {total_count:,}건")
print(f"상담내용 없는 데이터: {empty_count:,}건")
print(f"전체 대비 비율: {empty_ratio:.4f}%")

# 상담내용 없는 행 제거 후 정제 데이터 생성
df_clean = df[~empty_mask].copy().reset_index(drop=True)

print(f"정제 후 데이터: {len(df_clean):,}건")

전체 데이터: 26,572건
상담내용 없는 데이터: 9건
전체 대비 비율: 0.0339%
정제 후 데이터: 26,563건


# CHAPTER 1. 상담 업무별 데이터 필요성 분석

## 각 상담에서 다음 두 가지를 동시에 분석한다.

1. 어떤 데이터 Field가 상담 수행에 필요한가
2. 그 데이터가 어느 수준까지 구체적으로 요구되는가

구체화 수준:

- L0 CONTEXT_ONLY
  - 계좌, 카드, 대출 등 업무 대상만 언급

- L1 FIELD_REQUIRED
  - 계좌번호, 거래내역, 생년월일 등 필요한 Field가 명시됨

- L2 PARTIAL_VALUE
  - 일부 자리, 마스킹 값 등 부분 정보 수준

- L3 EXACT_VALUE
  - 실제 금액, 금리, 잔액, 날짜, 식별값 등 정확한 값이 요구됨

이를 통해 Field별로 KEEP / MASK / HMAC / TOKEN / DROP 설계에 필요한
Utility Requirement를 도출한다.

In [31]:
# 07. 데이터 Field 및 구체화 수준 규칙

FIELD_RULES = {

    "CUSTOMER_NAME": {
        "field": [r"고객명", r"성명", r"이름"],
        "partial": [r"[■●○*]{2,}", r"김\*", r"이\*", r"박\*"],
        "exact": []
    },

    "RESIDENT_ID": {
        "field": [r"주민등록번호", r"주민번호"],
        "partial": [
            r"\d{6}-?[1-4][*■●]{6}",
            r"\d{6}-?[*■●]{7}"
        ],
        "exact": [
            r"\d{6}-?[1-4]\d{6}"
        ]
    },

    "DATE_OF_BIRTH": {
        "field": [r"생년월일", r"출생일"],
        "partial": [],
        "exact": [
            r"\d{4}[.\-/년]\s*\d{1,2}[.\-/월]\s*\d{1,2}",
            r"\d{6}"
        ]
    },

    "PHONE_NUMBER": {
        "field": [r"전화번호", r"휴대폰\s*번호", r"연락처"],
        "partial": [
            r"01[016789]-?[*■●]{3,4}-?[*■●]{4}"
        ],
        "exact": [
            r"01[016789]-?\d{3,4}-?\d{4}"
        ]
    },

    "ACCOUNT_NUMBER": {
        "field": [r"계좌번호", r"계좌\s*번호"],
        "partial": [
            r"\d{2,6}[- ]?[*■●]{2,}[- ]?\d{2,6}",
            r"끝\s*\d+\s*자리"
        ],
        "exact": [
            r"\b\d{2,6}-\d{2,6}-\d{2,8}\b"
        ]
    },

    "CARD_NUMBER": {
        "field": [r"카드번호", r"카드\s*번호"],
        "partial": [
            r"\d{4}-?[*■●]{4}-?[*■●]{4}-?\d{4}",
            r"카드.*끝\s*4자리"
        ],
        "exact": [
            r"\b\d{4}-?\d{4}-?\d{4}-?\d{4}\b"
        ]
    },

    "TRANSACTION_HISTORY": {
        "field": [
            r"거래\s*내역", r"이체\s*내역",
            r"입출금\s*내역", r"결제\s*내역"
        ],
        "partial": [],
        "exact": []
    },

    "TRANSACTION_AMOUNT": {
        "field": [
            r"거래\s*금액", r"이체\s*금액",
            r"입금\s*금액", r"출금\s*금액",
            r"결제\s*금액"
        ],
        "partial": [],
        "exact": [
            r"\d[\d,]*\s*(?:원|만원|억원)"
        ]
    },

    "ACCOUNT_BALANCE": {
        "field": [r"잔액", r"계좌\s*잔액", r"예금\s*잔액"],
        "partial": [],
        "exact": [
            r"\d[\d,]*\s*(?:원|만원|억원)"
        ]
    },

    "INTEREST_RATE": {
        "field": [r"금리", r"이자율", r"연체\s*이자"],
        "partial": [],
        "exact": [
            r"\d+(?:\.\d+)?\s*%"
        ]
    },

    "LOAN_AMOUNT": {
        "field": [
            r"대출\s*금액", r"대출\s*잔액",
            r"대출\s*원금", r"상환\s*금액"
        ],
        "partial": [],
        "exact": [
            r"\d[\d,]*\s*(?:원|만원|억원)"
        ]
    },

    "DELINQUENCY_AMOUNT": {
        "field": [r"연체\s*금액", r"연체액"],
        "partial": [],
        "exact": [
            r"\d[\d,]*\s*(?:원|만원|억원)"
        ]
    },

    "DELINQUENCY_PERIOD": {
        "field": [r"연체\s*기간", r"연체\s*일수"],
        "partial": [],
        "exact": [
            r"\d+\s*(?:일|개월|년)"
        ]
    },

    "CREDIT_SCORE": {
        "field": [r"신용\s*점수", r"신용\s*등급"],
        "partial": [],
        "exact": [
            r"\d{3,4}\s*점"
        ]
    },

    "INCOME": {
        "field": [r"연\s*소득", r"월\s*소득", r"급여"],
        "partial": [],
        "exact": [
            r"\d[\d,]*\s*(?:원|만원|억원)"
        ]
    }
}

In [32]:
# 08. 상담별 Field 구체화 수준 태깅

import re

LEVEL_SCORE = {
    "CONTEXT_ONLY": 0,
    "FIELD_REQUIRED": 1,
    "PARTIAL_VALUE": 2,
    "EXACT_VALUE": 3
}


def analyze_field_requirement(text):

    text = str(text)
    result = []

    for field, rules in FIELD_RULES.items():

        field_hit = any(
            re.search(p, text, re.I)
            for p in rules["field"]
        )

        partial_hit = any(
            re.search(p, text, re.I)
            for p in rules["partial"]
        )

        exact_hit = any(
            re.search(p, text, re.I)
            for p in rules["exact"]
        )

        if not (field_hit or partial_hit or exact_hit):
            continue

        if exact_hit:
            level = "EXACT_VALUE"

        elif partial_hit:
            level = "PARTIAL_VALUE"

        else:
            level = "FIELD_REQUIRED"

        result.append({
            "field": field,
            "level": level,
            "score": LEVEL_SCORE[level]
        })

    return result


df_clean["data_requirements"] = (
    df_clean["consulting_content"]
    .apply(analyze_field_requirement)
)

In [33]:
# 09. 상담별 필요 Field와 구체화 수준 확인

df_clean["required_field_count"] = (
    df_clean["data_requirements"]
    .apply(len)
)

df_clean["max_specificity"] = (
    df_clean["data_requirements"]
    .apply(
        lambda x: max(
            [i["score"] for i in x],
            default=0
        )
    )
)

display(
    df_clean[
        [
            "consulting_content",
            "data_requirements",
            "required_field_count",
            "max_specificity"
        ]
    ].head(20)
)

,consulting_content,data_requirements,required_field_count,max_specificity
0,TX 안녕하세요 OO은행 고객센터입니다. 오늘 무엇을 도와드릴까요?\nRX 안녕하세...,"[{'field': 'CUSTOMER_NAME', 'level': 'PARTIAL_...",5,3
1,"고객은 최근 출금 거래의 ATM 위치와 수수료면제 여부, 해당계좌의 무단 사용 여부...",[],0,0
2,고객은 외화 계좌를 원화 계좌로 전환하고 카드 발급을 위해 필요한 절차와 최소 예치...,[],0,0
3,"고객은 오늘 입금된 내역과 금액을 확인하고 싶다고 문의했고, 상담원은 계좌 확인 후...",[],0,0
4,"고객은 최근 이체가 입금되지 않아 확인을 요청하고, 거래 중지 계좌의 자동이체 가능...","[{'field': 'TRANSACTION_HISTORY', 'level': 'FI...",1,1
5,"RX 안녕하세요, 저는 ■■■이고 현재 ○○은행에 거래하고 있는 고객입니다. 오늘 ...","[{'field': 'CUSTOMER_NAME', 'level': 'PARTIAL_...",5,3
6,"RX 네, 저는 ■■■■■입니다. 안녕하세요.\nRX 이게 지금 저기 뭐야 그 거기...","[{'field': 'CUSTOMER_NAME', 'level': 'PARTIAL_...",4,2
7,"TX 안녕하세요, OO은행입니다. 무엇을 도와드릴까요? 고객님께서 거래내역서를 원하...","[{'field': 'CUSTOMER_NAME', 'level': 'PARTIAL_...",3,2
8,TX 안녕하세요 OO은행 고객센터입니다. 무엇을 도와드릴까요?\nRX 안녕하세요. ...,"[{'field': 'CUSTOMER_NAME', 'level': 'PARTIAL_...",5,2
9,"TX 안녕하세요, 고객님. 무엇을 도와드릴까요?\nRX 안녕하세요. 제 통장에 자동...","[{'field': 'CUSTOMER_NAME', 'level': 'PARTIAL_...",5,2


In [34]:
# 10. Field 주변 문맥 기반 구체화 수준 판정

def get_context(text, match, window=50):
    start = max(0, match.start() - window)
    end = min(len(text), match.end() + window)
    return text[start:end]


def analyze_field_requirement_v2(text):

    text = str(text)
    results = []

    for field, rules in FIELD_RULES.items():

        best_level = None
        best_score = -1

        for field_pattern in rules["field"]:

            for match in re.finditer(
                field_pattern,
                text,
                flags=re.IGNORECASE
            ):

                context = get_context(
                    text,
                    match,
                    window=50
                )

                level = "FIELD_REQUIRED"

                # 정확값 확인
                if any(
                    re.search(p, context, re.I)
                    for p in rules["exact"]
                ):
                    level = "EXACT_VALUE"

                # 부분값 확인
                elif any(
                    re.search(p, context, re.I)
                    for p in rules["partial"]
                ):
                    level = "PARTIAL_VALUE"

                score = LEVEL_SCORE[level]

                if score > best_score:
                    best_level = level
                    best_score = score

        if best_level is not None:
            results.append({
                "field": field,
                "level": best_level,
                "score": best_score
            })

    return results


df_clean["data_requirements_v2"] = (
    df_clean["consulting_content"]
    .apply(analyze_field_requirement_v2)
)

In [35]:
# 11. Field × 구체화 수준 분석

requirements = (
    df_clean[["data_requirements_v2"]]
    .explode("data_requirements_v2")
    .dropna()
)

requirements["field"] = (
    requirements["data_requirements_v2"]
    .apply(lambda x: x["field"])
)

requirements["level"] = (
    requirements["data_requirements_v2"]
    .apply(lambda x: x["level"])
)

specificity_table = pd.crosstab(
    requirements["field"],
    requirements["level"]
)

specificity_ratio = (
    pd.crosstab(
        requirements["field"],
        requirements["level"],
        normalize="index"
    )
    * 100
).round(1)

display(specificity_table)
display(specificity_ratio)

level,EXACT_VALUE,FIELD_REQUIRED,PARTIAL_VALUE
field,,,
ACCOUNT_BALANCE,132,3813,0
ACCOUNT_NUMBER,3,8820,0
CARD_NUMBER,0,94,0
CREDIT_SCORE,4,904,0
CUSTOMER_NAME,0,2540,646
DATE_OF_BIRTH,16,6647,0
DELINQUENCY_AMOUNT,18,545,0
DELINQUENCY_PERIOD,29,146,0
INCOME,34,2269,0


level,EXACT_VALUE,FIELD_REQUIRED,PARTIAL_VALUE
field,,,
ACCOUNT_BALANCE,3.3,96.7,0.0
ACCOUNT_NUMBER,0.0,100.0,0.0
CARD_NUMBER,0.0,100.0,0.0
CREDIT_SCORE,0.4,99.6,0.0
CUSTOMER_NAME,0.0,79.7,20.3
DATE_OF_BIRTH,0.2,99.8,0.0
DELINQUENCY_AMOUNT,3.2,96.8,0.0
DELINQUENCY_PERIOD,16.6,83.4,0.0
INCOME,1.5,98.5,0.0


## 12. Field별 Value Requirement 분석

상담에서 특정 Field가 언급되는 것과
해당 Field의 실제 원문값이 필요한 것은 구분한다.

각 Field에 대해 전체 탐지 건수 중
부분값 또는 정확값까지 관측된 비율을 Value Requirement Ratio로 계산한다.

이를 통해 식별정보와 금융업무 수치정보 간
Utility Requirement의 차이를 비교하고,
Transform Treatment 설계 근거로 활용한다.

In [36]:
# 12. Field별 Value Requirement Ratio 계산

summary = specificity_table.copy()

for col in [
    "FIELD_REQUIRED",
    "PARTIAL_VALUE",
    "EXACT_VALUE"
]:
    if col not in summary.columns:
        summary[col] = 0

summary["TOTAL"] = (
    summary["FIELD_REQUIRED"]
    + summary["PARTIAL_VALUE"]
    + summary["EXACT_VALUE"]
)

summary["VALUE_REQUIRED"] = (
    summary["PARTIAL_VALUE"]
    + summary["EXACT_VALUE"]
)

summary["VALUE_REQUIREMENT_RATIO"] = (
    summary["VALUE_REQUIRED"]
    / summary["TOTAL"]
    * 100
).round(2)

summary["EXACT_REQUIREMENT_RATIO"] = (
    summary["EXACT_VALUE"]
    / summary["TOTAL"]
    * 100
).round(2)

display(
    summary[
        [
            "TOTAL",
            "FIELD_REQUIRED",
            "PARTIAL_VALUE",
            "EXACT_VALUE",
            "VALUE_REQUIREMENT_RATIO",
            "EXACT_REQUIREMENT_RATIO"
        ]
    ].sort_values(
        "VALUE_REQUIREMENT_RATIO",
        ascending=False
    )
)

level,TOTAL,FIELD_REQUIRED,PARTIAL_VALUE,EXACT_VALUE,VALUE_REQUIREMENT_RATIO,EXACT_REQUIREMENT_RATIO
field,,,,,,
CUSTOMER_NAME,3186,2540,646,0,20.28,0.00
DELINQUENCY_PERIOD,175,146,0,29,16.57,16.57
INTEREST_RATE,8138,6936,0,1202,14.77,14.77
TRANSACTION_AMOUNT,1508,1456,0,52,3.45,3.45
ACCOUNT_BALANCE,3945,3813,0,132,3.35,3.35
DELINQUENCY_AMOUNT,563,545,0,18,3.20,3.20
LOAN_AMOUNT,2019,1966,0,53,2.63,2.63
INCOME,2303,2269,0,34,1.48,1.48
PHONE_NUMBER,2662,2644,18,0,0.68,0.00


## 13. 데이터 유형별 Utility Requirement 재구성

Field마다 업무상 필요한 정보 수준의 의미가 다르므로
단일 EXACT_VALUE 기준을 모든 데이터에 동일하게 적용하지 않는다.

### Identifier
이름, 주민번호, 전화번호, 계좌번호, 카드번호
→ 전체 식별값 / 부분 식별값 / 식별값 불필요

### Numeric Value
금액, 잔액, 금리, 소득, 신용점수, 연체기간
→ 정확한 수치값 필요 여부

### Record / History
거래내역
→ 단순 내역 존재 여부가 아니라 기간·거래목록·세부내역 수준의 필요성

이를 기반으로 데이터 유형별 Transform 설계를 분리한다.

In [38]:
# 13. Field 유형 정의

field_type_map = {

    "CUSTOMER_NAME": "IDENTIFIER",
    "RESIDENT_ID": "IDENTIFIER",
    "DATE_OF_BIRTH": "IDENTIFIER",
    "PHONE_NUMBER": "IDENTIFIER",
    "ACCOUNT_NUMBER": "IDENTIFIER",
    "CARD_NUMBER": "IDENTIFIER",

    "INTEREST_RATE": "NUMERIC",
    "LOAN_AMOUNT": "NUMERIC",
    "DELINQUENCY_AMOUNT": "NUMERIC",
    "DELINQUENCY_PERIOD": "NUMERIC",
    "TRANSACTION_AMOUNT": "NUMERIC",
    "ACCOUNT_BALANCE": "NUMERIC",
    "CREDIT_SCORE": "NUMERIC",
    "INCOME": "NUMERIC",

    "TRANSACTION_HISTORY": "RECORD"
}

summary_analysis = summary.reset_index()

summary_analysis["field_type"] = (
    summary_analysis["field"]
    .map(field_type_map)
)

display(
    summary_analysis[
        [
            "field",
            "field_type",
            "TOTAL",
            "PARTIAL_VALUE",
            "EXACT_VALUE",
            "VALUE_REQUIREMENT_RATIO"
        ]
    ]
)

level,field,field_type,TOTAL,PARTIAL_VALUE,EXACT_VALUE,VALUE_REQUIREMENT_RATIO
0,ACCOUNT_BALANCE,NUMERIC,3945,0,132,3.35
1,ACCOUNT_NUMBER,IDENTIFIER,8823,0,3,0.03
2,CARD_NUMBER,IDENTIFIER,94,0,0,0.00
3,CREDIT_SCORE,NUMERIC,908,0,4,0.44
4,CUSTOMER_NAME,IDENTIFIER,3186,646,0,20.28
5,DATE_OF_BIRTH,IDENTIFIER,6663,0,16,0.24
6,DELINQUENCY_AMOUNT,NUMERIC,563,0,18,3.20
7,DELINQUENCY_PERIOD,NUMERIC,175,0,29,16.57
8,INCOME,NUMERIC,2303,0,34,1.48
9,INTEREST_RATE,NUMERIC,8138,0,1202,14.77


In [39]:
# 14. 분석결과 기반 1차 Treatment 후보

def suggest_treatment(row):

    field_type = row["field_type"]
    exact = row["EXACT_REQUIREMENT_RATIO"]
    value = row["VALUE_REQUIREMENT_RATIO"]

    if field_type == "IDENTIFIER":

        if exact < 1:
            return "MASK / HMAC / TOKEN 후보"
        else:
            return "조건부 원문 필요 검토"

    elif field_type == "NUMERIC":

        if exact >= 10:
            return "KEEP_EXACT 필요성 높음"
        elif exact > 0:
            return "조건부 KEEP_EXACT"
        else:
            return "최소화 가능"

    elif field_type == "RECORD":
        return "별도 상세수준 분석 필요"

    return "REVIEW"


summary_analysis["treatment_candidate"] = (
    summary_analysis.apply(
        suggest_treatment,
        axis=1
    )
)

display(
    summary_analysis[
        [
            "field",
            "field_type",
            "VALUE_REQUIREMENT_RATIO",
            "EXACT_REQUIREMENT_RATIO",
            "treatment_candidate"
        ]
    ].sort_values(
        ["field_type", "VALUE_REQUIREMENT_RATIO"],
        ascending=[True, False]
    )
)

level,field,field_type,VALUE_REQUIREMENT_RATIO,EXACT_REQUIREMENT_RATIO,treatment_candidate
4,CUSTOMER_NAME,IDENTIFIER,20.28,0.00,MASK / HMAC / TOKEN 후보
11,PHONE_NUMBER,IDENTIFIER,0.68,0.00,MASK / HMAC / TOKEN 후보
5,DATE_OF_BIRTH,IDENTIFIER,0.24,0.24,MASK / HMAC / TOKEN 후보
1,ACCOUNT_NUMBER,IDENTIFIER,0.03,0.03,MASK / HMAC / TOKEN 후보
2,CARD_NUMBER,IDENTIFIER,0.00,0.00,MASK / HMAC / TOKEN 후보
12,RESIDENT_ID,IDENTIFIER,0.00,0.00,MASK / HMAC / TOKEN 후보
7,DELINQUENCY_PERIOD,NUMERIC,16.57,16.57,KEEP_EXACT 필요성 높음
9,INTEREST_RATE,NUMERIC,14.77,14.77,KEEP_EXACT 필요성 높음
13,TRANSACTION_AMOUNT,NUMERIC,3.45,3.45,조건부 KEEP_EXACT
0,ACCOUNT_BALANCE,NUMERIC,3.35,3.35,조건부 KEEP_EXACT


## 17. 가설검정 ① 식별정보와 금융 수치정보의 정확값 요구 차이

### 연구가설

- H0: 식별정보(IDENTIFIER)와 금융 수치정보(NUMERIC)의 EXACT_VALUE 요구 비율은 동일하다.
- H1: 식별정보의 EXACT_VALUE 요구 비율이 금융 수치정보보다 낮다.

### 분석 목적

상담 업무에서 데이터 유형에 따라 정확 원문값의 필요성이 실제로 다른지 검증한다.

유의한 차이가 확인될 경우,
모든 Field를 동일한 방식으로 처리하는 것이 아니라
데이터 유형별로 Transform 수준을 차등 설계해야 한다는 근거로 사용한다.

In [40]:
# 17. 식별정보 vs 금융 수치정보 정확값 요구 비율 검정

from statsmodels.stats.proportion import proportions_ztest
from scipy.stats import chi2_contingency
import numpy as np

identifier = summary_analysis[
    summary_analysis["field_type"] == "IDENTIFIER"
]

numeric = summary_analysis[
    summary_analysis["field_type"] == "NUMERIC"
]

identifier_exact = identifier["EXACT_VALUE"].sum()
identifier_total = identifier["TOTAL"].sum()

numeric_exact = numeric["EXACT_VALUE"].sum()
numeric_total = numeric["TOTAL"].sum()

identifier_ratio = identifier_exact / identifier_total
numeric_ratio = numeric_exact / numeric_total

print("IDENTIFIER")
print(f"EXACT: {identifier_exact:,} / {identifier_total:,}")
print(f"비율: {identifier_ratio * 100:.3f}%")

print("\nNUMERIC")
print(f"EXACT: {numeric_exact:,} / {numeric_total:,}")
print(f"비율: {numeric_ratio * 100:.3f}%")

IDENTIFIER
EXACT: 19 / 22,274
비율: 0.085%

NUMERIC
EXACT: 1,524 / 19,559
비율: 7.792%


In [41]:
# 18. 단측 두 비율 검정

count = np.array([
    identifier_exact,
    numeric_exact
])

nobs = np.array([
    identifier_total,
    numeric_total
])

z_stat, p_value = proportions_ztest(
    count,
    nobs,
    alternative="smaller"
)

print(f"Z-statistic: {z_stat:.4f}")
print(f"P-value: {p_value:.6g}")

if p_value < 0.05:
    print(
        "\n결론: H0 기각\n"
        "식별정보의 정확값 요구 비율은 금융 수치정보보다 "
        "통계적으로 유의하게 낮다."
    )
else:
    print(
        "\n결론: H0 기각 불가\n"
        "식별정보와 금융 수치정보 간 정확값 요구 비율의 "
        "유의한 차이를 확인하지 못했다."
    )

Z-statistic: -41.7261
P-value: 0

결론: H0 기각
식별정보의 정확값 요구 비율은 금융 수치정보보다 통계적으로 유의하게 낮다.


In [42]:
# 19. 데이터 유형과 EXACT 요구 여부의 연관성 및 효과크기

contingency = np.array([
    [
        identifier_exact,
        identifier_total - identifier_exact
    ],
    [
        numeric_exact,
        numeric_total - numeric_exact
    ]
])

chi2, p, dof, expected = chi2_contingency(contingency)

n = contingency.sum()
phi = np.sqrt(chi2 / n)

print("Contingency Table")
print(contingency)

print(f"\nChi-square: {chi2:.4f}")
print(f"P-value: {p:.6g}")
print(f"Phi effect size: {phi:.4f}")

Contingency Table
[[   19 22255]
 [ 1524 18035]]

Chi-square: 1738.9002
P-value: 0
Phi effect size: 0.2039


## 18. 가설검정 ② 금융 수치 Field별 정확값 요구 차이

### 연구가설

- H0: 금융 수치 Field별 EXACT_VALUE 요구 비율은 동일하다.
- H1: 금융 수치 Field별 EXACT_VALUE 요구 비율에는 차이가 있다.

### 분석 목적

금융 수치정보라고 하더라도
금리·잔액·거래금액·대출금액 등의 업무적 Utility가 동일한지 검증한다.

차이가 존재한다면 NUMERIC Field 역시 일괄 KEEP_EXACT이 아니라
Field별 필요 수준에 따라 처리해야 한다.

In [43]:
# 20. 금융 수치 Field별 EXACT_VALUE 비율 차이 검정

numeric_test = summary_analysis[
    summary_analysis["field_type"] == "NUMERIC"
].copy()

numeric_test["NOT_EXACT"] = (
    numeric_test["TOTAL"]
    - numeric_test["EXACT_VALUE"]
)

contingency_numeric = numeric_test[
    [
        "EXACT_VALUE",
        "NOT_EXACT"
    ]
].to_numpy()

chi2, p, dof, expected = chi2_contingency(
    contingency_numeric
)

n = contingency_numeric.sum()

cramers_v = np.sqrt(
    chi2
    / (
        n
        * min(
            contingency_numeric.shape[0] - 1,
            contingency_numeric.shape[1] - 1
        )
    )
)

print(f"Chi-square: {chi2:.4f}")
print(f"df: {dof}")
print(f"P-value: {p:.6g}")
print(f"Cramer's V: {cramers_v:.4f}")

if p < 0.05:
    print(
        "\n결론: H0 기각\n"
        "금융 수치 Field별 EXACT_VALUE 요구 비율에는 "
        "통계적으로 유의한 차이가 있다."
    )
else:
    print(
        "\n결론: H0 기각 불가"
    )

Chi-square: 1006.2052
df: 7
P-value: 5.49592e-213
Cramer's V: 0.2268

결론: H0 기각
금융 수치 Field별 EXACT_VALUE 요구 비율에는 통계적으로 유의한 차이가 있다.


## 19. 상담 단위 가설검정

### 가설

- H0: 동일 상담 내에서 식별정보와 금융 수치정보의 EXACT_VALUE 요구 확률에는 차이가 없다.
- H1: 식별정보보다 금융 수치정보에서 EXACT_VALUE가 더 자주 요구된다.

동일한 상담 안에서 두 데이터 유형을 비교하므로
상담 단위의 paired test인 McNemar 검정을 사용한다.

In [45]:
# 21. 상담별 EXACT_VALUE 요구 여부 생성

identifier_fields = {
    "CUSTOMER_NAME",
    "RESIDENT_ID",
    "DATE_OF_BIRTH",
    "PHONE_NUMBER",
    "ACCOUNT_NUMBER",
    "CARD_NUMBER"
}

numeric_fields = {
    "INTEREST_RATE",
    "LOAN_AMOUNT",
    "DELINQUENCY_AMOUNT",
    "DELINQUENCY_PERIOD",
    "TRANSACTION_AMOUNT",
    "ACCOUNT_BALANCE",
    "CREDIT_SCORE",
    "INCOME"
}


def has_exact_requirement(requirements, field_set):

    return int(
        any(
            item["field"] in field_set
            and item["level"] == "EXACT_VALUE"
            for item in requirements
        )
    )


df_clean["identifier_exact"] = (
    df_clean["data_requirements_v2"]
    .apply(
        lambda x: has_exact_requirement(
            x,
            identifier_fields
        )
    )
)

df_clean["numeric_exact"] = (
    df_clean["data_requirements_v2"]
    .apply(
        lambda x: has_exact_requirement(
            x,
            numeric_fields
        )
    )
)

print(
    "식별정보 EXACT 상담:",
    df_clean["identifier_exact"].sum()
)

print(
    "금융수치 EXACT 상담:",
    df_clean["numeric_exact"].sum()
)

식별정보 EXACT 상담: 19
금융수치 EXACT 상담: 1403


In [46]:
# 22. McNemar 검정

from statsmodels.stats.contingency_tables import mcnemar
import pandas as pd

table = pd.crosstab(
    df_clean["identifier_exact"],
    df_clean["numeric_exact"]
)

table = table.reindex(
    index=[0, 1],
    columns=[0, 1],
    fill_value=0
)

display(table)

result = mcnemar(
    table,
    exact=True
)

print(f"McNemar statistic: {result.statistic}")
print(f"P-value: {result.pvalue:.6g}")

if result.pvalue < 0.05:
    print(
        "\n결론: H0 기각\n"
        "동일 상담 내에서도 금융 수치정보의 정확값 요구가 "
        "식별정보보다 유의하게 높다."
    )
else:
    print("\n결론: H0 기각 불가")

numeric_exact,0,1
identifier_exact,,
0,25142,1402
1,18,1


McNemar statistic: 18.0
P-value: 0

결론: H0 기각
동일 상담 내에서도 금융 수치정보의 정확값 요구가 식별정보보다 유의하게 높다.


## 19. 가설검정 종합 결과 — 데이터 유형별 Utility Requirement 차이

### 가설 ① 식별정보와 금융 수치정보의 정확값 요구 차이

식별정보와 금융 수치정보에서 정확한 원문값이 요구되는 비율이 동일한지 검정하였다.

- IDENTIFIER EXACT: 19 / 22,274
- NUMERIC EXACT: 1,524 / 19,559
- Z-statistic = -41.7261
- p < 0.001

**H0를 기각하였다.**

상담 데이터에서 식별정보의 정확값 요구 비율은
금융 수치정보보다 통계적으로 유의하게 낮았다.

---

### 가설 ② 동일 상담 내 데이터 유형별 정확값 요구 차이

동일한 상담 안에서도 식별정보와 금융 수치정보의
정확값 요구 여부가 다른지 McNemar 검정을 수행하였다.

- McNemar statistic = 18
- p < 0.001

**H0를 기각하였다.**

동일 상담 단위에서도 금융 수치정보가
식별정보보다 정확값으로 요구되는 경향이 유의하게 높았다.

이는 단순히 Field별 표본 수 차이에서 발생한 결과가 아니라,
실제 상담 단위에서도 데이터 유형에 따른 Utility Requirement 차이가 존재함을 보여준다.

---

### 가설 ③ 금융 수치 Field 내부의 정확값 요구 차이

금융 수치정보라고 하더라도
금리·잔액·거래금액·대출금액 등 모든 Field에서
정확값 요구 비율이 동일한지 검정하였다.

- χ² = 1006.2052
- df = 7
- p < 0.001
- Cramer's V = 0.2268

**H0를 기각하였다.**

금융 수치 Field별 EXACT_VALUE 요구 비율에는
통계적으로 유의한 차이가 존재하였다.

즉 금융 수치정보 역시 일괄적으로 KEEP_EXACT 처리하는 것이 아니라,
업무상 필요한 Utility 수준에 따라 Field별로 차등 처리할 필요가 있다.

---

## 종합 결론

세 가지 검정은 동일한 방향을 지지한다.

> **금융 상담에서 요구되는 데이터 Utility는 데이터 유형 및 Field별로 유의하게 다르다.**

따라서 Gateway의 Transform 정책은
모든 정보를 동일하게 마스킹하거나 동일하게 원문 유지하는 방식이 아니라,

**Field → 업무상 필요 수준 → Utility Requirement → Transform Treatment**

순으로 결정되어야 한다.

특히 상담 데이터에서는 식별정보의 정확 원문값 필요성이 매우 낮게 관측된 반면,
금리·연체기간 등 일부 금융 수치정보에서는 정확값 보존 필요성이 상대적으로 높게 나타났다.

따라서 다음 단계에서는 이 결과를 기반으로
각 Field를 `식별 유지 필요 / 관계 유지 필요 / 복원 필요 / 정확값 필요`로 구분한 뒤
MASK / HMAC / TOKEN / KEEP_EXACT Treatment와 연결한다.

## 20. Field별 Utility Requirement Matrix

상담 데이터 분석 결과를 기반으로 각 Field에 대해
Transform 기법을 바로 결정하지 않고 다음 업무 요구조건을 정의한다.

- Exact Requirement: 정확한 원문값이 필요한가
- Partial Requirement: 부분값만으로 업무가 가능한가
- Relationship Requirement: 동일 고객·계좌·거래 간 관계 유지가 필요한가
- Reversibility Requirement: 이후 원본 복원이 필요한가

이 4개 조건을 기반으로 최종 Transform Treatment를 결정한다.